benches/parser.rs

```diff
@@ -0,0 +1,60 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+use std::{fs, path::PathBuf, str::FromStr};
+
+fn benchmark_quil_corpus(c: &mut Criterion) {
+    from_corpus().iter().for_each(|cfg| {
+        c.bench_function(&cfg.name, |b| {
+            b.iter(|| {
+                let _ = quil_rs::Program::from_str(&cfg.program);
+            })
+        });
+    })
+}
+
+struct QuilBenchConfig {
+    name: String,
+    program: String,
+}
+
+fn from_corpus() -> Vec<QuilBenchConfig> {
+    const PATH_SRC: &str = "benches/quilc/tests/good-test-files";
+
+    // collect valid quil programs
+    let mut programs = vec![];
+    let mut corpus_dir = PathBuf::new();
+    PATH_SRC.split('/').for_each(|p| corpus_dir.push(p));
+    let dir = fs::read_dir(corpus_dir).expect("failed to locate quil corpus directory");
+
+    dir.filter_map(|res| {
+        if res.is_ok() {
+            Some(res.unwrap())
+        } else {
+            None
+        }
+    })
+    .for_each(|entry| {
+        if entry
+            .metadata()
+            .expect("failed to read file metadata")
+            .is_file()
+        {
+            let program =
+                fs::read_to_string(entry.path()).expect("failed to read quil program file");
+            let name = entry
+                .file_name()
+                .to_str()
+                .expect("bad filename")
+                .to_string();
+
+            // attempt to parse the quil once, ignoring unparsable input (only benchmark parsable code)
+            if quil_rs::Program::from_str(&program).is_ok() {
+                programs.push(QuilBenchConfig { name, program });
+            }
+        }
+    });
```

Comment:
Move the for_each predicate into a filter. Spacing looks weird in suggestion-mode, I might have messed something up 😬.

```suggestion
    .filter(|entry| {
        entry
            .metadata()
            .expect("failed to read file metadata")
            .is_file()
    })
    .for_each(|entry| {
        let program =
            fs::read_to_string(entry.path()).expect("failed to read quil program file");
        let name = entry
            .file_name()
            .to_str()
            .expect("bad filename")
            .to_string();

        // attempt to parse the quil once, ignoring unparsable input (only benchmark parsable code)
        if quil_rs::Program::from_str(&program).is_ok() {
            programs.push(QuilBenchConfig { name, program });
        }
    });
```